homeassistant/components/modem_callerid/config_flow.py

```diff
@@ -0,0 +1,149 @@
+
+from .const import DEFAULT_NAME, DOMAIN, EXCEPTIONS
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema({"name": str, "device": str})
+
+
+def _generate_unique_id(port: Any) -> str:
+    """Generate unique id from usb attributes."""
+    return f"{port.vid}:{port.pid}_{port.serial_number}_{port.manufacturer}_{port.description}"
+
+
+class PhoneModemFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Phone Modem."""
+
+    def __init__(self) -> None:
+        """Set up flow instance."""
+        self._device: str | None = None
+
+    async def async_step_usb(self, discovery_info: dict[str, str]) -> FlowResult:
+        """Handle USB Discovery."""
+        vid = discovery_info["vid"]
+        pid = discovery_info["pid"]
+        serial_number = discovery_info["serial_number"]
+        device = discovery_info["device"]
+        manufacturer = discovery_info["manufacturer"]
+        description = discovery_info["description"]
+
+        dev_path = await self.hass.async_add_executor_job(usb.get_serial_by_id, device)
+        unique_id = f"{vid}:{pid}_{serial_number}_{manufacturer}_{description}"
+        if await self.validate_input(dev_path=dev_path, unique_id=unique_id) is None:
+            self._device = dev_path
+            return await self.async_step_usb_confirm()
+        return self.async_abort(reason="cannot_connect")
+
+    async def async_step_usb_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle USB Discovery confirmation."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title=user_input.get(CONF_NAME, DEFAULT_NAME),
+                data={CONF_DEVICE: self._device},
+            )
+        self._set_confirm_only()
+        return self.async_show_form(step_id="usb_confirm")
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initiated by the user."""
+        if self._async_in_progress():
+            return self.async_abort(reason="already_in_progress")
+        ports = await self.hass.async_add_executor_job(serial.tools.list_ports.comports)
+        if self.source == config_entries.SOURCE_IMPORT and user_input is not None:
+            for port in ports:
+                if port.device == user_input[CONF_DEVICE]:
+                    if (
+                        await self.validate_input(
+                            user_input=user_input,
+                            dev_path=port.device,
+                            unique_id=_generate_unique_id(port),
+                        )
+                        is None
+                    ):
+                        return self.async_create_entry(
+                            title=user_input.get(CONF_NAME, DEFAULT_NAME),
+                            data={CONF_DEVICE: port.device},
+                        )
```

Comment:
It looks like it might be a bit cleaner if this was moved to `async_step_import`